### PR TITLE
Type stabilize offset_data

### PR DESCRIPTION
**Pull request by wsmoses**
cc @christophernhill

### CODE REVIEW

**Comment by navidcy** [comment]
@wsmoses, can you write a small description in the first comment summarizing what this PR do or fixes?

**Comment by glwagner** [comment]
I think these minor changes allow / encourage the compiler to infer the output type of `offset_data` --- is that right? As far as I can tell it does not change the type that's outputted by `offset_data`. But I could be wrong.

Once we confirm the intent we can merge (it could make sense also to add a comment to the code so that this doesn't get reverted for some reason in the future). @wsmoses let us know if you'd like to tag a patch release as well.

**Comment by wsmoses** [comment]
@glwagner yes indeed, the intent is to allow the compiler to infer the return (and internals) of offset_data. It does not change the result or result types.

File: src/Grids/new_data.jl [hunk 1]
```diff
@@ -32,12 +32,15 @@ offset_indices(::Nothing, topo, N, H, ::UnitRange) = 1:1
 instantiate(T::Type) = T()
 instantiate(t) = t
 
-function offset_data(underlying_data::AbstractArray, loc, topo, N, H, indices=default_indices(length(loc)))
+function offset_data(underlying_data::A, loc, topo, N, H, indices::T=default_indices(length(loc))) where {A<:AbstractArray, T}
```
Comment (glwagner) [inline src/Grids/new_data.jl:35]:
```suggestion
# The type parameter for indices helps / encourages the compiler to fully type infer `offset_data`
function offset_data(underlying_data::A, loc, topo, N, H, indices::T=default_indices(length(loc))) where {A<:AbstractArray, T}
```

File: src/Grids/new_data.jl [hunk 2]
```diff
@@ -32,12 +32,15 @@ offset_indices(::Nothing, topo, N, H, ::UnitRange) = 1:1
...
     loc = map(instantiate, loc)
     topo = map(instantiate, topo)
     ii = map(offset_indices, loc, topo, N, H, indices)
     # Add extra indices for arrays of higher dimension than loc, topo, etc.
-    extra_ii = Tuple(axes(underlying_data, d) for d in length(ii)+1:ndims(underlying_data))
+    extra_ii = ntuple(Val(ndims(underlying_data)-length(ii))) do i
```
Comment (glwagner) [inline src/Grids/new_data.jl:40]:
```suggestion
    # Use the "`ntuple` trick" so the compiler can infer the type of `extra_ii`
    extra_ii = ntuple(Val(ndims(underlying_data)-length(ii))) do i
```

**Comment by glwagner** [comment]
Nice, let's merge ! Patch release or no? There will probably be a patch release before long anyways.

**Comment by navidcy** [comment]
We'll have a minor release soon with #3125, right? So no need to bother here with patch?

**Comment by navidcy** [comment]
@glwagner your comments implies you approve but you didn't formally approve. Is there something left to be done here?

**Review by glwagner — APPROVED** [review]

**Comment by glwagner** [comment]
> @glwagner your comments implies you approve but you didn't formally approve. Is there something left to be done here?

Nothing left. Merged.

**Pull request merged**